prefect_aws/ecs.py

```diff
@@ -0,0 +1,981 @@
+            )
+
+    def _report_task_run_creation_failure(self, task_run: dict, exc: Exception) -> None:
+        """
+        Wrap common AWS task run creation failures with nicer user-facing messages.
+        """
+        # AWS generates exception types at runtime so they must be captured a bit
+        # differently than normal.
+        if "ClusterNotFoundException" in str(exc):
+            cluster = task_run.get("cluster", "default")
+            raise RuntimeError(
+                f"Failed to run ECS task, cluster {cluster!r} not found. "
+                "Confirm that the cluster is configured in your region."
+            ) from exc
+        elif "No Container Instances" in str(exc) and self.launch_type == "EC2":
+            cluster = task_run.get("cluster", "default")
+            raise RuntimeError(
+                f"Failed to run ECS task, cluster {cluster!r} does not appear to "
+                "have any container instances associated with it. Confirm that you "
+                "have EC2 container instances available."
+            ) from exc
+        elif (
+            "failed to validate logger args" in str(exc)
+            and "AccessDeniedException" in str(exc)
+            and self.configure_cloudwatch_logs
+        ):
+
+            raise RuntimeError(
+                f"Failed to run ECS task, the attached execution role does not appear "
+                "to have sufficient permissions. Ensure that the execution role "
+                f"{self.execution_role!r} has permissions logs:CreateLogStream, "
+                "logs:CreateLogGroup, and logs:PutLogEvents."
+            )
+        else:
+            raise
+
+    def _watch_task_run(
+        self,
+        task_arn: str,
+        cluster_arn: str,
+        ecs_client: _ECSClient,
+        current_status: str = "UNKNOWN",
+        until_status: str = None,
+        timeout: int = None,
+    ) -> Generator[None, None, dict]:
+        """
+        Watches an ECS task run by querying every `poll_interval` seconds. After each
+        query, the retrieved task is yielded. This function returns when the task run
+        reaches a STOPPED status or the provided `until_status`.
+
+        Emits a log each time the status changes.
+        """
+        last_status = status = current_status
+        t0 = time.time()
+        while status != until_status:
+            task = ecs_client.describe_tasks(tasks=[task_arn], cluster=cluster_arn)[
+                "tasks"
+            ][0]
+
+            status = task["lastStatus"]
+            if status != last_status:
+                self.logger.info(f"{self.log_prefix}: Status is {status}.")
+
+            yield task
+
+            # No point in continuing if the status is final
+            if status == "STOPPED":
+                break
+
+            last_status = status
+            elapsed_time = time.time() - t0
+            if timeout is not None and elapsed_time > timeout:
+                raise TimeoutError(
```

Comment:
My understanding is that `TimeoutError` is used for when a [a system function times out at the system level](https://docs.python.org/3/library/exceptions.html#TimeoutError). Should we raise a custom exception for this type of timeout?

Reply:
I'm not sure. Despite that documentation, I see this timeout error around. I'm generally hesitant to introduce new exception types outside the core library. I could be convinced though. cc @bunchesofdonald

Reply:
I like using built-in exceptions when feasible. I do agree that TimeoutError is a weird case though, it's a subclass of OSError, which isn't applicable here. I think if we want to use something else here we should create a core exception and replace the other ~5 usages of it in Prefect (on a glance none seem to be about system calls), but that could be a larger headache.

Reply:
I think often in Prefect we're just reraising the lower-level error from AnyIO/asyncio which _must_ be an actual `TimeoutError` or things will not tear down correctly. Hmm...

Reply:
Could we raise a [`RuntimError`](https://docs.python.org/3/library/exceptions.html#RuntimeError) here instead since it seems like it's the most applicable of the  built-in exceptions?

Reply:
I guess.. that's not really accurate either but we can improve this later.

Reply:
```suggestion
                raise RuntimeError(
```